Gems/Terrain/Code/Source/TerrainRenderer/TerrainDetailMaterialManager.h

```diff
@@ -0,0 +1,208 @@
+
+    private:
+        
+        using MaterialInstance = AZ::Data::Instance<AZ::RPI::Material>;
+        static constexpr auto InvalidImageIndex = AZ::Render::BindlessImageArrayHandler::InvalidImageIndex;
+
+        enum DetailTextureFlags : uint32_t
+        {
+            UseTextureBaseColor =  0b0000'0000'0000'0000'0000'0000'0000'0001,
+            UseTextureNormal =     0b0000'0000'0000'0000'0000'0000'0000'0010,
+            UseTextureMetallic =   0b0000'0000'0000'0000'0000'0000'0000'0100,
+            UseTextureRoughness =  0b0000'0000'0000'0000'0000'0000'0000'1000,
+            UseTextureOcclusion =  0b0000'0000'0000'0000'0000'0000'0001'0000,
+            UseTextureHeight =     0b0000'0000'0000'0000'0000'0000'0010'0000,
+            UseTextureSpecularF0 = 0b0000'0000'0000'0000'0000'0000'0100'0000,
+
+            FlipNormalX =          0b0000'0000'0000'0001'0000'0000'0000'0000,
+            FlipNormalY =          0b0000'0000'0000'0010'0000'0000'0000'0000,
+
+            BlendModeMask =        0b0000'0000'0000'1100'0000'0000'0000'0000,
+            BlendModeLerp =        0b0000'0000'0000'0000'0000'0000'0000'0000,
+            BlendModeLinearLight = 0b0000'0000'0000'0100'0000'0000'0000'0000,
+            BlendModeMultiply =    0b0000'0000'0000'1000'0000'0000'0000'0000,
+            BlendModeOverlay =     0b0000'0000'0000'1100'0000'0000'0000'0000,
+        };
+        
+        struct DetailMaterialShaderData
+        {
+            // Uv
+            AZStd::array<float, 12> m_uvTransform
+            {
+                1.0, 0.0, 0.0, 0.0,
+                0.0, 1.0, 0.0, 0.0,
+                0.0, 0.0, 1.0, 0.0,
+            };
+
+            float m_baseColorRed{ 1.0f };
+            float m_baseColorGreen{ 1.0f };
+            float m_baseColorBlue{ 1.0f };
+
+            // Factor / Scale / Bias for input textures
+            float m_baseColorFactor{ 1.0f };
+
+            float m_normalFactor{ 1.0f };
+            float m_metalFactor{ 1.0f };
+            float m_roughnessScale{ 1.0f };
+            float m_roughnessBias{ 0.0f };
+
+            float m_specularF0Factor{ 1.0f };
+            float m_occlusionFactor{ 1.0f };
+            float m_heightFactor{ 1.0f };
+            float m_heightOffset{ 0.0f };
+
+            float m_heightBlendFactor{ 0.5f };
+
+            // Flags
+            DetailTextureFlags m_flags{ 0 };
+
+            // Image indices
+            uint16_t m_colorImageIndex{ InvalidImageIndex };
+            uint16_t m_normalImageIndex{ InvalidImageIndex };
+            uint16_t m_roughnessImageIndex{ InvalidImageIndex };
+            uint16_t m_metalnessImageIndex{ InvalidImageIndex };
+
+            uint16_t m_specularF0ImageIndex{ InvalidImageIndex };
+            uint16_t m_occlusionImageIndex{ InvalidImageIndex };
+            uint16_t m_heightImageIndex{ InvalidImageIndex };
+
+            // 16 byte aligned
+            uint16_t m_padding1;
+            uint32_t m_padding2;
+            uint32_t m_padding3;
```

Comment:
Suggestion:  Might want to consider adding a static_assert to validate the sizeof(DetailMaterialShaderData) % 16 == 0, to help enforce that this stays correct over time.

Reply:
yup, adding this to all the shader structs reflected in c++, I like it.